package/kedro_viz/models/flowchart.py

```diff
@@ -364,100 +416,174 @@ def outputs(self) -> Set[str]:
         )
 
 
-@dataclass
 class TaskNodeMetadata(GraphNodeMetadata):
-    """Represent the metadata of a TaskNode"""
+    """Represent the metadata of a TaskNode
+
+    Args:
+        task_node (TaskNode): Task node to which this metadata belongs to.
+
+    Raises:
+        AssertionError: If task_node is not supplied during instantiation
+    """
 
-    # the source code of the node's function
-    code: Optional[str] = field(init=False)
+    task_node: TaskNode
 
-    # path to the file where the node is defined
-    filepath: Optional[str] = field(init=False)
+    # Source code of the node's function
+    code: Optional[str]
+
+    # Path to the file where the node is defined
+    filepath: Optional[str]
+
+    parameters: Optional[Dict] = Field(
+        None, description="The parameters of the node, if available"
+    )
+    run_command: Optional[str] = Field(
+        None, description="The command to run the pipeline to this node"
+    )
 
-    # parameters of the node, if available
-    parameters: Optional[Dict] = field(init=False, default=None)
+    inputs: Optional[List[str]] = Field(None, description="The inputs to the TaskNode")
+    outputs: Optional[List[str]] = Field(
+        None, description="The outputs from the TaskNode"
+    )
 
-    # command to run the pipeline to this node
-    run_command: Optional[str] = field(init=False, default=None)
+    @root_validator(pre=True)
+    def check_task_node_exists(cls, values):
+        assert "task_node" in values
+        cls.set_task_and_kedro_node(values["task_node"])
+        return values
 
-    inputs: List[str] = field(init=False)
+    @classmethod
+    def set_task_and_kedro_node(cls, task_node):
```

Comment:
nit - I got a bit confused with the function name. maybe we just call it set_task_node